libc/sem.cc

```diff
@@ -76,3 +104,72 @@ int sem_trywait(sem_t* s)
         return libc_error(EAGAIN);
     return 0;
 }
+
+static std::unordered_map<std::string, indirect_semaphore> named_semaphores;
+static mutex named_semaphores_mutex;
+
+OSV_LIBC_API
+sem_t *sem_open(const char *name, int oflag, mode_t mode, unsigned int value)
+{
+    SCOPE_LOCK(named_semaphores_mutex);
+    auto iter = named_semaphores.find(std::string(name));
+    
+    if (iter != named_semaphores.end()) {
+        //opening already named semaphore
+        if (oflag & O_EXCL) {
+            errno = EEXIST;
+            return SEM_FAILED;
+        }
+
+        iter->second.get()->add_reference();
```

Comment:
I don't think you need to do second.get()->add_reference() - you can simply do the shorter second->add_reference(). Doesn't that work?